### PR TITLE
Allow sending a specific page of a pagination list and add a test plugin

### PR DESCRIPTION
**Pull request by parlough**
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/1622) | SpongeCommon

Accomplishes part of https://github.com/SpongePowered/SpongeCommon/pull/1305

Also cleans up some messages and code and makes it so a pagination list defaults to an empty list if no contents are specified.

### CODE REVIEW

**Review by JBYoshi — APPROVED** [review]

File: src/main/java/org/spongepowered/common/service/pagination/SpongePaginationList.java [hunk 1]
```diff
@@ -100,9 +101,9 @@ public int getLinesPerPage() {
     }
 
     @Override
-    public void sendTo(final MessageReceiver receiver) {
-        checkNotNull(this.contents, "contents");
-        checkNotNull(receiver, "source");
+    public void sendTo(final MessageReceiver receiver, int page) {
+        checkNotNull(this.contents, "The contents of the pagination list cannot be null!");
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/common/service/pagination/SpongePaginationList.java]: This is a final field. This should be validated in the constructor.

**Review by ST-DDT — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/service/pagination/SpongePaginationList.java (hunk 1 above)
Comment (parlough) [reply src/main/java/org/spongepowered/common/service/pagination/SpongePaginationList.java]: It's validated in the builder build method, so I just removed this check.

File: src/main/java/org/spongepowered/common/service/pagination/SpongePaginationBuilder.java [hunk 2]
```diff
@@ -105,7 +106,9 @@ public SpongePaginationBuilder(SpongePaginationService service) {
 
     @Override
     public PaginationList build() {
-        checkNotNull(this.contents, "contents");
+        if (this.contents == null) {
+            this.contents = ImmutableList.of();
+        }
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/common/service/pagination/SpongePaginationBuilder.java]: IMO its more likely to be an error to not specify the contents. So this should throw an ISException. But it does not really matter.

**Review by ST-DDT — APPROVED** [review]

File: src/main/java/org/spongepowered/common/service/pagination/SpongePaginationBuilder.java (hunk 2 above)
Comment (parlough) [reply src/main/java/org/spongepowered/common/service/pagination/SpongePaginationBuilder.java]: I have made this change

File: src/main/java/org/spongepowered/common/service/pagination/IterablePagination.java [hunk 3]
```diff
@@ -63,7 +63,7 @@ public IterablePagination(Supplier<Optional<MessageReceiver>> src, PaginationCal
         }
 
         if (page <= this.lastPage) {
-            throw new CommandException(t("Cannot go backward in an IterablePagination"));
+            throw new CommandException(t("You cannot go backwards in an IterablePagination."));
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/pagination/IterablePagination.java]: Perhaps phrase this exception message a different way? This sounds weird.

File: src/main/java/org/spongepowered/common/service/pagination/ListPagination.java [hunk 4]
```diff
@@ -85,7 +85,7 @@ public ListPagination(Supplier<Optional<MessageReceiver>> src, PaginationCalcula
         } else if (page < 1) {
             throw new CommandException(t("Page %s does not exist!", page));
         } else if (page > this.pages.size()) {
-            throw new CommandException(t("Page %s is too high", page));
+            throw new CommandException(t("Page %s is too high!", page));
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/pagination/ListPagination.java]: Include the maximum page?
`"Page %s is greater than %s."`?

File: src/main/java/org/spongepowered/common/service/pagination/SpongePaginationBuilder.java [hunk 5]
```diff
@@ -91,7 +92,7 @@ public SpongePaginationBuilder(SpongePaginationService service) {
 
     @Override
     public PaginationList.Builder padding(Text padding) {
-        checkNotNull(padding, "padding");
+        checkNotNull(padding, "The padding cannot be null!");
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/pagination/SpongePaginationBuilder.java:96]: unnecessary

File: src/main/java/org/spongepowered/common/service/pagination/SpongePaginationBuilder.java [hunk 6]
```diff
@@ -105,7 +106,9 @@ public SpongePaginationBuilder(SpongePaginationService service) {
 
     @Override
     public PaginationList build() {
-        checkNotNull(this.contents, "contents");
+        if (this.contents == null) {
+            throw new IllegalStateException("The contents of the pagination list cannot be null!");
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/pagination/SpongePaginationBuilder.java]: use `checkState`

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/service/pagination/IterablePagination.java (hunk 3 above)
Comment (parlough) [reply src/main/java/org/spongepowered/common/service/pagination/IterablePagination.java]: `You cannot go to a previous page in an IterablePagination.`?

Comment (kashike) [reply src/main/java/org/spongepowered/common/service/pagination/IterablePagination.java]: Sure.

**Pull request closed without merging**